File: sample/cpp/18_mini_language_interpreter.cpp
#include "runtime/cpp/pytra/built_in/py_runtime.h"

#include "pytra/std/dataclasses.h"
#include "pytra/std/time.h"

struct Token {
    str kind;
    str text;
    int64 pos;
    int64 number_value;
    
    Token(str kind, str text, int64 pos, int64 number_value) : kind(kind), text(text), pos(pos), number_value(number_value) {
    }
    
};

struct ExprNode {
    str kind;
    int64 value;
    str name;
    str op;
    int64 left;
    int64 right;
    int64 kind_tag;
    int64 op_tag;
    
    ExprNode(str kind, int64 value, str name, str op, int64 left, int64 right, int64 kind_tag, int64 op_tag) : kind(kind), value(value), name(name), op(op), left(left), right(right), kind_tag(kind_tag), op_tag(op_tag) {
    }
    
};

struct StmtNode {
    str kind;
    str name;
    int64 expr_index;
    int64 kind_tag;
    
    StmtNode(str kind, str name, int64 expr_index, int64 kind_tag) : kind(kind), name(name), expr_index(expr_index), kind_tag(kind_tag) {
    }
    
};

list<Token> tokenize(const list<str>& lines) {
    dict<str, int64> single_char_token_tags = dict<str, int64>{{"+", 1}, {"-", 2}, {"*", 3}, {"/", 4}, {"(", 5}, {")", 6}, {"=", 7}};
    list<str> single_char_token_kinds = list<str>{"PLUS", "MINUS", "STAR", "SLASH", "LPAREN", "RPAREN", "EQUAL"};
    list<Token> tokens = {};
    for (const auto& [line_index, source] : py_enumerate(lines)) {
        int64 i = 0;
        int64 n = py_len(source);
        while (i < n) {
            str ch = source[i];
            
            if (ch == " ") {
                i++;
                continue;
            }
            int64 single_tag = py_to<int64>(single_char_token_tags.get(ch, 0));
            if (single_tag > 0) {
                tokens.append(Token(Token(single_char_token_kinds[single_tag - 1], ch, i, 0)));
                i++;
                continue;
            }
            if (ch.isdigit()) {
                int64 start = i;
                while ((i < n) && (source[i].isdigit())) {
                    i++;
                }
                str text = py_slice(source, start, i);
                tokens.append(Token(Token("NUMBER", text, start, py_to_int64(text))));
                continue;
            }
            if ((ch.isalpha()) || (ch == "_")) {
                int64 start = i;
                while ((i < n) && (((source[i].isalpha()) || (source[i] == "_")) || (source[i].isdigit()))) {
                    i++;
                }
                str text = py_slice(source, start, i);
                if (text == "let") {
                    tokens.append(Token(Token("LET", text, start, 0)));
                } else if (text == "print") {
                    tokens.append(Token(Token("PRINT", text, start, 0)));
                } else {
                    tokens.append(Token(Token("IDENT", text, start, 0)));
                }
                continue;
            }
            throw ::std::runtime_error("tokenize error at line=" + ::std::to_string(line_index) + " pos=" + ::std::to_string(i) + " ch=" + ch);
        }
        tokens.append(Token(Token("NEWLINE", "", n, 0)));
    }
    tokens.append(Token(Token("EOF", "", py_len(lines), 0)));
    return tokens;
}

struct Parser : public PyObj {
    list<ExprNode> expr_nodes;
    int64 pos;
    list<Token> tokens;
    PYTRA_DECLARE_CLASS_TYPE(PYTRA_TID_OBJECT);
    
    list<ExprNode> new_expr_nodes() {
        return list<ExprNode>{};
    }
    Parser(const list<Token>& tokens) {
        this->tokens = tokens;
        this->pos = 0;
        this->expr_nodes = this->new_expr_nodes();
    }
    Token current_token() {
        return this->tokens[this->pos];
    }
    Token previous_token() {
        return this->tokens[this->pos - 1];
    }
    str peek_kind() {
        return this->current_token().kind;
    }
    bool match(const str& kind) {
        if (this->peek_kind() == kind) {
            this->pos++;
            return true;
        }
        return false;
    }
    Token expect(const str& kind) {
        Token token = this->current_token();
        if (token.kind != kind)
            throw ::std::runtime_error("parse error at pos=" + py_to_string(token.pos) + ", expected=" + kind + ", got=" + token.kind);
        this->pos++;
        return token;
    }
    void skip_newlines() {
        while (this->match("NEWLINE")) {
            ;
        }
    }
    int64 add_expr(const ExprNode& node) {
        this->expr_nodes.append(node);
        return py_len(this->expr_nodes) - 1;
    }
    list<StmtNode> parse_program() {
        list<StmtNode> stmts = {};
        this->skip_newlines();
        while (this->peek_kind() != "EOF") {
            StmtNode stmt = this->parse_stmt();
            stmts.append(stmt);
            this->skip_newlines();
        }
        return stmts;
    }
    StmtNode parse_stmt() {
        if (this->match("LET")) {
            str let_name = this->expect("IDENT").text;
            this->expect("EQUAL");
            int64 let_expr_index = this->parse_expr();
            return StmtNode("let", let_name, let_expr_index, 1);
        }
        if (this->match("PRINT")) {
            int64 print_expr_index = this->parse_expr();
            return StmtNode("print", "", print_expr_index, 3);
        }
        str assign_name = this->expect("IDENT").text;
        this->expect("EQUAL");
        int64 assign_expr_index = this->parse_expr();
        return StmtNode("assign", assign_name, assign_expr_index, 2);
    }
    int64 parse_expr() {
        return this->parse_add();
    }
    int64 parse_add() {
        int64 left = this->parse_mul();
        while (true) {
            if (this->match("PLUS")) {
                int64 right = this->parse_mul();
                left = this->add_expr(ExprNode("bin", 0, "", "+", left, right, 3, 1));
                continue;
            }
            if (this->match("MINUS")) {
                int64 right = this->parse_mul();
                left = this->add_expr(ExprNode("bin", 0, "", "-", left, right, 3, 2));
                continue;
            }
            break;
        }
        return left;
    }
    int64 parse_mul() {
        int64 left = this->parse_unary();
        while (true) {
            if (this->match("STAR")) {
                int64 right = this->parse_unary();
                left = this->add_expr(ExprNode("bin", 0, "", "*", left, right, 3, 3));
                continue;
            }
            if (this->match("SLASH")) {
                int64 right = this->parse_unary();
                left = this->add_expr(ExprNode("bin", 0, "", "/", left, right, 3, 4));
                continue;
            }
            break;
        }
        return left;
    }
    int64 parse_unary() {
        if (this->match("MINUS")) {
            int64 child = this->parse_unary();
            return this->add_expr(ExprNode("neg", 0, "", "", child, -1, 4, 0));
        }
        return this->parse_primary();
    }
    int64 parse_primary() {
        if (this->match("NUMBER")) {
            Token token_num = this->previous_token();
            return this->add_expr(ExprNode("lit", token_num.number_value, "", "", -1, -1, 1, 0));
        }
        if (this->match("IDENT")) {
            Token token_ident = this->previous_token();
            return this->add_expr(ExprNode("var", 0, token_ident.text, "", -1, -1, 2, 0));
        }
        if (this->match("LPAREN")) {
            int64 expr_index = this->parse_expr();
            this->expect("RPAREN");
            return expr_index;
        }
        Token t = this->current_token();
        throw ::std::runtime_error("primary parse error at pos=" + py_to_string(t.pos) + " got=" + t.kind);
    }
};

int64 eval_expr(int64 expr_index, const list<ExprNode>& expr_nodes, const dict<str, int64>& env) {
    ExprNode node = expr_nodes[expr_index];
    
    if (node.kind_tag == 1)
        return node.value;
    if (node.kind_tag == 2) {
        if (!(py_contains(env, node.name)))
            throw ::std::runtime_error("undefined variable: " + node.name);
        return py_dict_get(env, node.name);
    }
    if (node.kind_tag == 4)
        return -eval_expr(node.left, expr_nodes, env);
    if (node.kind_tag == 3) {
        int64 lhs = eval_expr(node.left, expr_nodes, env);
        int64 rhs = eval_expr(node.right, expr_nodes, env);
        if (node.op_tag == 1)
            return lhs + rhs;
        if (node.op_tag == 2)
            return lhs - rhs;
        if (node.op_tag == 3)
            return lhs * rhs;
        if (node.op_tag == 4) {
            if (rhs == 0)
                throw ::std::runtime_error("division by zero");
            return lhs / rhs;
        }
        throw ::std::runtime_error("unknown operator: " + node.op);
    }
    throw ::std::runtime_error("unknown node kind: " + node.kind);
}

int64 execute(const list<StmtNode>& stmts, const list<ExprNode>& expr_nodes, bool trace) {
    dict<str, int64> env = {};
    int64 checksum = 0;
    int64 printed = 0;
    
    for (StmtNode stmt : stmts) {
        if (stmt.kind_tag == 1) {
            env[stmt.name] = eval_expr(stmt.expr_index, expr_nodes, env);
            continue;
        }
        if (stmt.kind_tag == 2) {
            if (!(py_contains(env, stmt.name)))
                throw ::std::runtime_error("assign to undefined variable: " + stmt.name);
            env[stmt.name] = eval_expr(stmt.expr_index, expr_nodes, env);
            continue;
        }
        int64 value = eval_expr(stmt.expr_index, expr_nodes, env);
        if (trace)
            py_print(value);
        int64 norm = value % 1000000007;
        if (norm < 0)
            norm += 1000000007;
        checksum = (checksum * 131 + norm) % 1000000007;
        printed++;
    }
    if (trace)
        py_print("printed:", printed);
    return checksum;
}

list<str> build_benchmark_source(int64 var_count, int64 loops) {
    list<str> lines = {};
    
    // Declare initial variables.
    lines.reserve((var_count <= 0) ? 0 : var_count);
    for (int64 i = 0; i < var_count; ++i)
        lines.append(str("let v" + ::std::to_string(i) + " = " + ::std::to_string(i + 1)));
    // Force evaluation of many arithmetic expressions.
    for (int64 i = 0; i < loops; ++i) {
        int64 x = i % var_count;
        int64 y = (i + 3) % var_count;
        int64 c1 = i % 7 + 1;
        int64 c2 = i % 11 + 2;
        lines.append(str("v" + ::std::to_string(x) + " = (v" + ::std::to_string(x) + " * " + ::std::to_string(c1) + " + v" + ::std::to_string(y) + " + 10000) / " + ::std::to_string(c2)));
        if (i % 97 == 0)
            lines.append(str("print v" + ::std::to_string(x)));
    }
    // Print final values together.
    lines.append(str("print (v0 + v1 + v2 + v3)"));
    return lines;
}

void run_demo() {
    list<str> demo_lines = {};
    demo_lines.append(str("let a = 10"));
    demo_lines.append(str("let b = 3"));
    demo_lines.append(str("a = (a + b) * 2"));
    demo_lines.append(str("print a"));
    demo_lines.append(str("print a / b"));
    
    list<Token> tokens = tokenize(demo_lines);
    rc<Parser> parser = ::rc_new<Parser>(tokens);
    list<StmtNode> stmts = parser->parse_program();
    int64 checksum = execute(stmts, parser->expr_nodes, true);
    py_print("demo_checksum:", checksum);
}

void run_benchmark() {
    list<str> source_lines = build_benchmark_source(32, 120000);
    float64 start = pytra::std::time::perf_counter();
    list<Token> tokens = tokenize(source_lines);
    rc<Parser> parser = ::rc_new<Parser>(tokens);
    list<StmtNode> stmts = parser->parse_program();
    int64 checksum = execute(stmts, parser->expr_nodes, false);
    float64 elapsed = pytra::std::time::perf_counter() - start;
    
    py_print("token_count:", py_len(tokens));
    py_print("expr_count:", py_len(parser->expr_nodes));
    py_print("stmt_count:", py_len(stmts));
    py_print("checksum:", checksum);
    py_print("elapsed_sec:", elapsed);
}

void __pytra_main() {
    run_demo();
    run_benchmark();
}

int main(int argc, char** argv) {
    pytra_configure_from_argv(argc, argv);
    __pytra_main();
    return 0;
}

File: sample/zig/18_mini_language_interpreter.zig
const std = @import("std");
const pytra = @import("built_in/py_runtime.zig");
const time = @import("std/time.zig");
const perf_counter = time.perf_counter;

const Token = struct {
    kind: []const u8,
    text: []const u8,
    pos: i64,
    number_value: i64,
};

const ExprNode = struct {
    kind: []const u8,
    value: i64,
    name: []const u8,
    op: []const u8,
    left: i64,
    right: i64,
    kind_tag: i64,
    op_tag: i64,
};

const StmtNode = struct {
    kind: []const u8,
    name: []const u8,
    expr_index: i64,
    kind_tag: i64,
};

fn tokenize(lines: pytra.Obj) pytra.Obj {
    const single_char_token_tags: std.StringHashMap(i64) = pytra.make_str_dict_from(i64, &[_][]const u8{ "+", "-", "*", "/", "(", ")", "=" }, &[_]i64{ 1, 2, 3, 4, 5, 6, 7 });
    const single_char_token_kinds: pytra.Obj = pytra.list_from([]const u8, &[_][]const u8{ "PLUS", "MINUS", "STAR", "SLASH", "LPAREN", "RPAREN", "EQUAL" });
    const tokens: pytra.Obj = pytra.list_from(*Token, &[_]*Token{  });
    var __enum_idx_1: i64 = 0;
    for (pytra.list_items(lines, []const u8)) |source| {
        const line_index: i64 = __enum_idx_1;
        var i: i64 = 0;
        const n: i64 = @as(i64, @intCast(source.len));
        while ((i < n)) {
            const ch: []const u8 = pytra.str_index(source, i);
            
            if (std.mem.eql(u8, ch, " ")) {
                i += 1;
                continue;
            }
            const single_tag: i64 = pytra.dict_get_default(i64, single_char_token_tags, ch, 0);
            if ((single_tag > 0)) {
                pytra.list_append(tokens, *Token, pytra.make_object(Token, Token{ .kind = pytra.list_get(single_char_token_kinds, []const u8, (single_tag - 1)), .text = ch, .pos = i, .number_value = 0 }));
                i += 1;
                continue;
            }
            var start: i64 = undefined;
            var text: []const u8 = undefined;
            if (pytra.char_isdigit(ch)) {
                start = i;
                while (((i < n) and pytra.char_isdigit(pytra.str_index(source, i)))) {
                    i += 1;
                }
                text = pytra.str_slice(source, start, i);
                pytra.list_append(tokens, *Token, pytra.make_object(Token, Token{ .kind = "NUMBER", .text = text, .pos = start, .number_value = pytra.str_to_int(text) }));
                continue;
            }
            if ((pytra.char_isalpha(ch) or std.mem.eql(u8, ch, "_"))) {
                start = i;
                while (((i < n) and ((pytra.char_isalpha(pytra.str_index(source, i)) or std.mem.eql(u8, pytra.str_index(source, i), "_")) or pytra.char_isdigit(pytra.str_index(source, i))))) {
                    i += 1;
                }
                text = pytra.str_slice(source, start, i);
                if (std.mem.eql(u8, text, "let")) {
                    pytra.list_append(tokens, *Token, pytra.make_object(Token, Token{ .kind = "LET", .text = text, .pos = start, .number_value = 0 }));
                } else {
                    if (std.mem.eql(u8, text, "print")) {
                        pytra.list_append(tokens, *Token, pytra.make_object(Token, Token{ .kind = "PRINT", .text = text, .pos = start, .number_value = 0 }));
                    } else {
                        pytra.list_append(tokens, *Token, pytra.make_object(Token, Token{ .kind = "IDENT", .text = text, .pos = start, .number_value = 0 }));
                    }
                }
                continue;
            }
            @panic(pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat("tokenize error at line=", pytra.to_str(line_index)), " pos="), pytra.to_str(i)), " ch="), ch));
        }
        pytra.list_append(tokens, *Token, pytra.make_object(Token, Token{ .kind = "NEWLINE", .text = "", .pos = n, .number_value = 0 }));
        __enum_idx_1 += 1;
    }
    pytra.list_append(tokens, *Token, pytra.make_object(Token, Token{ .kind = "EOF", .text = "", .pos = pytra.list_len(lines, []const u8), .number_value = 0 }));
    return tokens;
}

const Parser = struct {
    tokens: pytra.Obj = undefined,
    pos: i64 = undefined,
    expr_nodes: pytra.Obj = undefined,
    pub fn new_expr_nodes(_: *const Parser) pytra.Obj {
        return pytra.list_from(pytra.Obj, &[_]pytra.Obj{  });
    }
    
    pub fn init(tokens: pytra.Obj) Parser {
        var self: Parser = undefined;
        self.tokens = tokens;
        self.pos = 0;
        self.expr_nodes = self.new_expr_nodes();
        return self;
    }
    
    pub fn current_token(self: *const Parser) *Token {
        return pytra.list_get(self.tokens, *Token, self.pos);
    }
    
    pub fn previous_token(self: *const Parser) *Token {
        return pytra.list_get(self.tokens, *Token, (self.pos - 1));
    }
    
    pub fn peek_kind(self: *const Parser) []const u8 {
        return self.current_token().kind;
    }
    
    pub fn match(self: *Parser, kind: []const u8) bool {
        if (std.mem.eql(u8, self.peek_kind(), kind)) {
            self.pos += 1;
            return true;
        }
        return false;
    }
    
    pub fn expect(self: *Parser, kind: []const u8) *Token {
        const token: *Token = self.current_token();
        if (!std.mem.eql(u8, token.kind, kind)) {
            @panic(pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat("parse error at pos=", pytra.to_str(token.pos)), ", expected="), kind), ", got="), token.kind));
        }
        self.pos += 1;
        return token;
    }
    
    pub fn skip_newlines(self: *Parser) void {
        while (self.match("NEWLINE")) {
            // pass
        }
    }
    
    pub fn add_expr(self: *Parser, node: *ExprNode) i64 {
        pytra.list_append(self.expr_nodes, *ExprNode, node);
        return (pytra.list_len(self.expr_nodes, *ExprNode) - 1);
    }
    
    pub fn parse_program(self: *Parser) pytra.Obj {
        const stmts: pytra.Obj = pytra.list_from(*StmtNode, &[_]*StmtNode{  });
        self.skip_newlines();
        while (!std.mem.eql(u8, self.peek_kind(), "EOF")) {
            const stmt: *StmtNode = self.parse_stmt();
            pytra.list_append(stmts, *StmtNode, stmt);
            self.skip_newlines();
        }
        return stmts;
    }
    
    pub fn parse_stmt(self: *Parser) *StmtNode {
        if (self.match("LET")) {
            const let_name: []const u8 = self.expect("IDENT").text;
            _ = self.expect("EQUAL");
            const let_expr_index: i64 = self.parse_expr();
            return pytra.make_object(StmtNode, StmtNode{ .kind = "let", .name = let_name, .expr_index = let_expr_index, .kind_tag = 1 });
        }
        if (self.match("PRINT")) {
            const print_expr_index: i64 = self.parse_expr();
            return pytra.make_object(StmtNode, StmtNode{ .kind = "print", .name = "", .expr_index = print_expr_index, .kind_tag = 3 });
        }
        const assign_name: []const u8 = self.expect("IDENT").text;
        _ = self.expect("EQUAL");
        const assign_expr_index: i64 = self.parse_expr();
        return pytra.make_object(StmtNode, StmtNode{ .kind = "assign", .name = assign_name, .expr_index = assign_expr_index, .kind_tag = 2 });
    }
    
    pub fn parse_expr(self: *Parser) i64 {
        return self.parse_add();
    }
    
    pub fn parse_add(self: *Parser) i64 {
        var left: i64 = self.parse_mul();
        while (true) {
            var right: i64 = undefined;
            if (self.match("PLUS")) {
                right = self.parse_mul();
                left = self.add_expr(pytra.make_object(ExprNode, ExprNode{ .kind = "bin", .value = 0, .name = "", .op = "+", .left = left, .right = right, .kind_tag = 3, .op_tag = 1 }));
                continue;
            }
            if (self.match("MINUS")) {
                right = self.parse_mul();
                left = self.add_expr(pytra.make_object(ExprNode, ExprNode{ .kind = "bin", .value = 0, .name = "", .op = "-", .left = left, .right = right, .kind_tag = 3, .op_tag = 2 }));
                continue;
            }
            break;
        }
        return left;
    }
    
    pub fn parse_mul(self: *Parser) i64 {
        var left: i64 = self.parse_unary();
        while (true) {
            var right: i64 = undefined;
            if (self.match("STAR")) {
                right = self.parse_unary();
                left = self.add_expr(pytra.make_object(ExprNode, ExprNode{ .kind = "bin", .value = 0, .name = "", .op = "*", .left = left, .right = right, .kind_tag = 3, .op_tag = 3 }));
                continue;
            }
            if (self.match("SLASH")) {
                right = self.parse_unary();
                left = self.add_expr(pytra.make_object(ExprNode, ExprNode{ .kind = "bin", .value = 0, .name = "", .op = "/", .left = left, .right = right, .kind_tag = 3, .op_tag = 4 }));
                continue;
            }
            break;
        }
        return left;
    }
    
    pub fn parse_unary(self: *Parser) i64 {
        if (self.match("MINUS")) {
            const child: i64 = self.parse_unary();
            return self.add_expr(pytra.make_object(ExprNode, ExprNode{ .kind = "neg", .value = 0, .name = "", .op = "", .left = child, .right = -1, .kind_tag = 4, .op_tag = 0 }));
        }
        return self.parse_primary();
    }
    
    pub fn parse_primary(self: *Parser) i64 {
        if (self.match("NUMBER")) {
            const token_num: *Token = self.previous_token();
            return self.add_expr(pytra.make_object(ExprNode, ExprNode{ .kind = "lit", .value = token_num.number_value, .name = "", .op = "", .left = -1, .right = -1, .kind_tag = 1, .op_tag = 0 }));
        }
        if (self.match("IDENT")) {
            const token_ident: *Token = self.previous_token();
            return self.add_expr(pytra.make_object(ExprNode, ExprNode{ .kind = "var", .value = 0, .name = token_ident.text, .op = "", .left = -1, .right = -1, .kind_tag = 2, .op_tag = 0 }));
        }
        if (self.match("LPAREN")) {
            const expr_index: i64 = self.parse_expr();
            _ = self.expect("RPAREN");
            return expr_index;
        }
        const t: *Token = self.current_token();
        @panic(pytra.str_concat(pytra.str_concat(pytra.str_concat("primary parse error at pos=", pytra.to_str(t.pos)), " got="), t.kind));
    }
    
};

fn eval_expr(expr_index: i64, expr_nodes: pytra.Obj, env: std.StringHashMap(i64)) i64 {
    const node: *ExprNode = pytra.list_get(expr_nodes, *ExprNode, expr_index);
    
    if ((node.kind_tag == 1)) {
        return node.value;
    }
    if ((node.kind_tag == 2)) {
        if (!pytra.contains(env, node.name)) {
            @panic(pytra.str_concat("undefined variable: ", node.name));
        }
        return pytra.dict_get_default(i64, env, node.name, 0);
    }
    if ((node.kind_tag == 4)) {
        return -eval_expr(node.left, expr_nodes, env);
    }
    if ((node.kind_tag == 3)) {
        const lhs: i64 = eval_expr(node.left, expr_nodes, env);
        const rhs: i64 = eval_expr(node.right, expr_nodes, env);
        if ((node.op_tag == 1)) {
            return (lhs + rhs);
        }
        if ((node.op_tag == 2)) {
            return (lhs - rhs);
        }
        if ((node.op_tag == 3)) {
            return (lhs * rhs);
        }
        if ((node.op_tag == 4)) {
            if ((rhs == 0)) {
                @panic("division by zero");
            }
            return @divFloor(lhs, rhs);
        }
        @panic(pytra.str_concat("unknown operator: ", node.op));
    }
    @panic(pytra.str_concat("unknown node kind: ", node.kind));
}

fn execute(stmts: pytra.Obj, expr_nodes: pytra.Obj, trace: bool) i64 {
    var env: std.StringHashMap(i64) = pytra.make_str_dict(i64);
    var checksum: i64 = 0;
    var printed: i64 = 0;
    
    for (pytra.list_items(stmts, *StmtNode)) |stmt| {
        if ((stmt.kind_tag == 1)) {
            env.put(stmt.name, eval_expr(stmt.expr_index, expr_nodes, env)) catch {};
            continue;
        }
        if ((stmt.kind_tag == 2)) {
            if (!pytra.contains(env, stmt.name)) {
                @panic(pytra.str_concat("assign to undefined variable: ", stmt.name));
            }
            env.put(stmt.name, eval_expr(stmt.expr_index, expr_nodes, env)) catch {};
            continue;
        }
        const value: i64 = eval_expr(stmt.expr_index, expr_nodes, env);
        if (trace) {
            pytra.print(value);
        }
        var norm: i64 = @mod(value, 1000000007);
        if ((norm < 0)) {
            norm += 1000000007;
        }
        checksum = @mod(((checksum * 131) + norm), 1000000007);
        printed += 1;
    }
    if (trace) {
        pytra.print2("printed:", printed);
    }
    return checksum;
}

fn build_benchmark_source(var_count: i64, loops: i64) pytra.Obj {
    const lines: pytra.Obj = pytra.list_from([]const u8, &[_][]const u8{  });
    var i: i64 = undefined;
    
    // Declare initial variables.
    i = 0;
    while (i < var_count) : (i += 1) {
        pytra.list_append(lines, []const u8, pytra.str_concat(pytra.str_concat(pytra.str_concat("let v", pytra.to_str(i)), " = "), pytra.to_str((i + 1))));
    }
    // Force evaluation of many arithmetic expressions.
    i = 0;
    while (i < loops) : (i += 1) {
        const x: i64 = @mod(i, var_count);
        const y: i64 = @mod((i + 3), var_count);
        const c1: i64 = (@mod(i, 7) + 1);
        const c2: i64 = (@mod(i, 11) + 2);
        pytra.list_append(lines, []const u8, pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat(pytra.str_concat("v", pytra.to_str(x)), " = (v"), pytra.to_str(x)), " * "), pytra.to_str(c1)), " + v"), pytra.to_str(y)), " + 10000) / "), pytra.to_str(c2)));
        if ((@mod(i, 97) == 0)) {
            pytra.list_append(lines, []const u8, pytra.str_concat("print v", pytra.to_str(x)));
        }
    }
    // Print final values together.
    pytra.list_append(lines, []const u8, "print (v0 + v1 + v2 + v3)");
    return lines;
}

fn run_demo() void {
    const demo_lines: pytra.Obj = pytra.list_from([]const u8, &[_][]const u8{  });
    pytra.list_append(demo_lines, []const u8, "let a = 10");
    pytra.list_append(demo_lines, []const u8, "let b = 3");
    pytra.list_append(demo_lines, []const u8, "a = (a + b) * 2");
    pytra.list_append(demo_lines, []const u8, "print a");
    pytra.list_append(demo_lines, []const u8, "print a / b");
    
    const tokens: pytra.Obj = tokenize(demo_lines);
    const parser: *Parser = pytra.make_object(Parser, Parser.init(tokens));
    const stmts: pytra.Obj = parser.parse_program();
    const checksum: i64 = execute(stmts, parser.expr_nodes, true);
    pytra.print2("demo_checksum:", checksum);
}

fn run_benchmark() void {
    const source_lines: pytra.Obj = build_benchmark_source(32, 120000);
    const start: f64 = pytra.perf_counter();
    const tokens: pytra.Obj = tokenize(source_lines);
    const parser: *Parser = pytra.make_object(Parser, Parser.init(tokens));
    const stmts: pytra.Obj = parser.parse_program();
    const checksum: i64 = execute(stmts, parser.expr_nodes, false);
    const elapsed: f64 = (pytra.perf_counter() - start);
    
    pytra.print2("token_count:", pytra.list_len(tokens, *Token));
    pytra.print2("expr_count:", pytra.list_len(parser.expr_nodes, i64));
    pytra.print2("stmt_count:", pytra.list_len(stmts, *StmtNode));
    pytra.print2("checksum:", checksum);
    pytra.print2("elapsed_sec:", elapsed);
}

fn __pytra_main() void {
    run_demo();
    run_benchmark();
}

pub fn main() void {
    __pytra_main();
}
